summoner-cli/examples/full-batteries/.github/workflows/ci.yml

```diff
@@ -20,28 +21,92 @@ jobs:
           - "8.6.5"
           - "8.8.4"
           - "8.10.1"
+        exclude:
+          - os: macOS-latest
+            ghc: 8.0.2
+          - os: macOS-latest
+            ghc: 8.2.2
+          - os: macOS-latest
+            ghc: 8.4.4
+          - os: macOS-latest
+            ghc: 8.6.5
+          - os: macOS-latest
+            ghc: 8.8.4
+          - os: windows-latest
+            ghc: 8.0.2
+          - os: windows-latest
+            ghc: 8.2.2
+          - os: windows-latest
+            ghc: 8.4.4
+          - os: windows-latest
+            ghc: 8.6.5
+          - os: windows-latest
+            ghc: 8.8.4
 
     steps:
     - uses: actions/checkout@v2
       if: github.event.action == 'opened' || github.event.action == 'synchronize' || github.event.ref == 'refs/heads/master'
 
-    - uses: actions/setup-haskell@v1
+    - uses: actions/setup-haskell@v1.1.1
+      id: setup-haskell-cabal
       name: Setup Haskell
       with:
         ghc-version: ${{ matrix.ghc }}
         cabal-version: ${{ matrix.cabal }}
 
+    - name: Freeze
+      run: |
+        cabal v2-freeze
+
     - uses: actions/cache@v1
       name: Cache ~/.cabal/store
       with:
-        path: ~/.cabal/store
-        key: ${{ runner.os }}-${{ matrix.ghc }}-cabal
+        path: ${{ steps.setup-haskell-cabal.outputs.cabal-store }}
+        key: ${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('cabal.project.freeze') }}
+
+    - name: Build
+      run: |
+        cabal v2-configure --enable-tests --enable-benchmarks
+        cabal v2-build all
+
+    - name: Test
+      run: |
+        cabal v2-test all
+
+  stack:
+    name: stack / ghc ${{ matrix.ghc }}
+    runs-on: ubuntu-latest
```

Comment:
@chshersh I went with your approach from https://kodimensional.dev/github-actions and used only Linux, but I don't understand the reason behind it. Why not do the same thing as for cabal in terms of versions(all for Linux, latest for rest) and oses?

Reply:
The current approach described in my blog post optimizes for the low number of builds and rapid development cycles, so you don't need to wait hours for the CI to finish. Usually, for Haskell projects it's enough to do the following:

* Test at least once with each GHC version
* Test at least once on each OS
* Test at least once with Cabal
* Test at least once with Stack

Since testing multiple GHC versions with Stack is more awkward than with Cabal and because Stack build cache is much bigger than the Cabal cache, we prefer Cabal by default with different combinations of GHC and OS, and Stack is there only to make sure that your `stack.yaml` is up-to-date.

Yes, there are extremely rare cases when some combinations of GHC versions and OS may not work (and since GHC is broken on Windows most of the time, you will get a lot of failed CI builds on Windows). But testing everything on everything is usually a waste of time for most projects.